scalajs:  compiler crash when overriding Promise.then

## Compiler version

3.0.0

## Minimized code

```Scala
package x

import scalajs.*

class MyPromise[T]() extends js.Promise[T]( (resolve, reject) => ???) {

    override def `then`[S](
      onFulfilled: js.Function1[T, S | js.Thenable[S]],
      onRejected: js.UndefOr[js.Function1[scala.Any, S | js.Thenable[S]]] = js.undefined): js.Promise[S] =
        js.Promise( (resolve, reject) => ??? )

    override def `then`[S >: T](
      onFulfilled: Unit,
      onRejected: js.UndefOr[js.Function1[scala.Any, S | js.Thenable[S]]]): js.Promise[S] =
        `then`[S]( x => x , onRejected)

    override def `catch`[S >: T](
      onRejected: js.UndefOr[js.Function1[scala.Any, S | js.Thenable[S]]] = js.undefined): js.Promise[S] =
        `then`[S]( x => x , onRejected )

}

```

## Output (click arrow to expand)
<details>

```scala
[info] welcome to sbt 1.5.2 (N/A Java 15.0.2)
[info] loading global plugins from /Users/rssh/.sbt/1.0/plugins
[info] loading settings for project scalajs-then-override-build from plugins.sbt ...
[info] loading project definition from /Users/rssh/tests/dotty/scalajs-then-override/project
[info] loading settings for project root from build.sbt ...
[info] set current project to test (in build file:/Users/rssh/tests/dotty/scalajs-then-override/)
[info] Executing in batch mode. For better performance use sbt's shell
[info] compiling 1 Scala source to /Users/rssh/tests/dotty/scalajs-then-override/target/scala-3.0.0/classes ...
[info] exception occurred while compiling /Users/rssh/tests/dotty/scalajs-then-override/src/main/scala/x/X.scala
java.lang.AssertionError: assertion failed: found overloaded default getter def then$default$2(): scala.scalajs.js.| <and> 
  def then$default$2(): scala.runtime.BoxedUnit while compiling /Users/rssh/tests/dotty/scalajs-then-override/src/main/scala/x/X.scala
[error] ## Exception when compiling 1 sources to /Users/rssh/tests/dotty/scalajs-then-override/target/scala-3.0.0/classes
[error] java.lang.AssertionError: assertion failed: found overloaded default getter def then$default$2(): scala.scalajs.js.| <and> 
[error]   def then$default$2(): scala.runtime.BoxedUnit
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.backend.sjs.JSExportsGen.genCallDefaultGetter(JSExportsGen.scala:744)
[error] dotty.tools.backend.sjs.JSExportsGen.genPrepareArgs$$anonfun$2(JSExportsGen.scala:718)
[error] scala.collection.Iterator$$anon$9.next(Iterator.scala:575)
[error] scala.collection.mutable.Growable.addAll(Growable.scala:62)
[error] scala.collection.mutable.Growable.addAll$(Growable.scala:57)
[error] scala.collection.immutable.VectorBuilder.addAll(Vector.scala:1565)
[error] scala.collection.immutable.Vector$.from(Vector.scala:1349)
[error] scala.collection.immutable.Vector$.from(Vector.scala:34)
[error] scala.collection.IterableOps$WithFilter.map(Iterable.scala:884)
[error] dotty.tools.backend.sjs.JSExportsGen.genPrepareArgs(JSExportsGen.scala:731)
[error] dotty.tools.backend.sjs.JSExportsGen.genApplyForSingleExportedNonJSSuperCall(JSExportsGen.scala:681)
[error] dotty.tools.backend.sjs.JSExportsGen.genApplyForSingleExported(JSExportsGen.scala:636)
[error] dotty.tools.backend.sjs.JSExportsGen.genExportSameArgcRec(JSExportsGen.scala:540)
[error] dotty.tools.backend.sjs.JSExportsGen.genExportSameArgc(JSExportsGen.scala:516)
[error] dotty.tools.backend.sjs.JSExportsGen.$anonfun$28(JSExportsGen.scala:472)
[error] scala.collection.Iterator$$anon$9.next(Iterator.scala:575)
[error] scala.collection.immutable.List.prependedAll(List.scala:153)
[error] scala.collection.immutable.List$.from(List.scala:651)
[error] scala.collection.immutable.List$.from(List.scala:648)
[error] scala.collection.IterableOps$WithFilter.map(Iterable.scala:884)
[error] dotty.tools.backend.sjs.JSExportsGen.genExportMethodMultiAlts(JSExportsGen.scala:474)
[error] dotty.tools.backend.sjs.JSExportsGen.genExportMethod(JSExportsGen.scala:441)
[error] dotty.tools.backend.sjs.JSExportsGen.genMemberExportOrDispatcher$$anonfun$1(JSExportsGen.scala:333)
[error] dotty.tools.backend.sjs.JSCodeGen.withNewLocalNameScope$$anonfun$1(JSCodeGen.scala:117)
[error] dotty.tools.backend.sjs.ScopedVar$.withScopedVars(ScopedVar.scala:35)
[error] dotty.tools.backend.sjs.JSCodeGen.withNewLocalNameScope(JSCodeGen.scala:118)
[error] dotty.tools.backend.sjs.JSExportsGen.genMemberExportOrDispatcher(JSExportsGen.scala:334)
[error] dotty.tools.backend.sjs.JSExportsGen.genJSClassDispatcher(JSExportsGen.scala:323)
[error] dotty.tools.backend.sjs.JSExportsGen.genJSClassDispatchers$$anonfun$1(JSExportsGen.scala:295)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.backend.sjs.JSExportsGen.genJSClassDispatchers(JSExportsGen.scala:295)
[error] dotty.tools.backend.sjs.JSCodeGen.genNonNativeJSClass(JSCodeGen.scala:537)
[error] dotty.tools.backend.sjs.JSCodeGen.genCompilationUnit$$anonfun$6$$anonfun$1(JSCodeGen.scala:211)
[error] dotty.tools.backend.sjs.ScopedVar$.withScopedVars(ScopedVar.scala:35)
[error] dotty.tools.backend.sjs.JSCodeGen.genCompilationUnit$$anonfun$2(JSCodeGen.scala:221)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.backend.sjs.JSCodeGen.genCompilationUnit(JSCodeGen.scala:223)
[error] dotty.tools.backend.sjs.JSCodeGen.run(JSCodeGen.scala:152)
[error] dotty.tools.backend.sjs.GenSJSIR.run(GenSJSIR.scala:15)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:303)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:304)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:205)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:215)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:223)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:230)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:166)
[error] dotty.tools.dotc.Run.compile(Run.scala:150)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
```
</details>

--- a/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala
+++ b/compiler/src/dotty/tools/backend/sjs/JSCodeGen.scala
@@ -18,6 +18,7 @@ import Decorators._
 import Flags._
 import dotty.tools.dotc.ast.Trees._
 import Names._
+import NameKinds.DefaultGetterName
 import Types._
 import Symbols._
 import Denotations._
@@ -1068,6 +1069,14 @@ class JSCodeGen()(using genCtx: Context) {
             OptimizerHints.empty, None))
       } else if (sym.isJSNativeCtorDefaultParam) {
         // #11592
+        None
+      } else if (sym.is(Bridge) && sym.name.is(DefaultGetterName) && currentClassSym.isNonNativeJSClass) {
+        /* #12572 Bridges for default accessors in non-native JS classes must not be emitted,
+         * because they call another default accessor, making their entire body an
+         * <undefined-param> that cannot be eliminated.
+         * Such methods are never called anyway, because they are filtered out in
+         * JSExportsGen.defaultGetterDenot().
+         */
         None
       } else /*if (sym.isClassConstructor && isHijackedBoxedClass(sym.owner)) {
         None

--- a/compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala
+++ b/compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala
@@ -774,7 +774,7 @@ final class JSExportsGen(jsCodeGen: JSCodeGen)(using Context) {
     else sym.owner
 
   private def defaultGetterDenot(targetSym: Symbol, sym: Symbol, paramIndex: Int): Denotation =
-    targetSym.info.member(DefaultGetterName(sym.name.asTermName, paramIndex))
+    targetSym.info.memberBasedOnFlags(DefaultGetterName(sym.name.asTermName, paramIndex), excluded = Bridge)
 
   private def defaultGetterDenot(sym: Symbol, paramIndex: Int): Denotation =
     defaultGetterDenot(targetSymForDefaultGetter(sym), sym, paramIndex)

--- a/tests/sjs-junit/test/org/scalajs/testsuite/compiler/RegressionTestScala3.scala
+++ b/tests/sjs-junit/test/org/scalajs/testsuite/compiler/RegressionTestScala3.scala
@@ -35,6 +35,10 @@ class RegressionTestScala3 {
 
     assertEquals(5, foo(5)(4))
   }
+
+  @Test def defaultAccessorBridgesIssue12572(): Unit = {
+    new MyPromiseIssue12572[Int](5)
+  }
 }
 
 object RegressionTestScala3 {
@@ -52,6 +56,25 @@ object RegressionTestScala3 {
   @JSGlobal("RangeError")
   class RangeErrorIssue11592(msg: String = js.native) extends js.Object {
     val message: String = js.native
+  }
+
+  class MyPromiseIssue12572[T](t: T) extends js.Promise[T]((resolve, reject) => resolve(t)) {
+    override def `then`[S](
+        onFulfilled: js.Function1[T, S | js.Thenable[S]],
+        onRejected: js.UndefOr[js.Function1[scala.Any, S | js.Thenable[S]]] = js.undefined): js.Promise[S] = {
+      ???
+    }
+
+    override def `then`[S >: T](
+        onFulfilled: Unit,
+        onRejected: js.UndefOr[js.Function1[scala.Any, S | js.Thenable[S]]]): js.Promise[S] = {
+      ???
+    }
+
+    override def `catch`[S >: T](
+        onRejected: js.UndefOr[js.Function1[scala.Any, S | js.Thenable[S]]] = js.undefined): js.Promise[S] = {
+      ???
+    }
   }
 }
 

